### ISSUE TITLE
[no-invalid-void-type] Should void be allowed in generic arguments to function calls?

### ISSUE THREAD

**Issue by jtbandes** [typescript-eslint/typescript-eslint#3612]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

[This rule](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-invalid-void-type.md) allows `void` as a generic argument in types like `Promise<void>`, but disallows it in function generic arguments like `makePromise<void>()`. This makes it difficult to adopt the rule because generic function arguments often end up being used in generic type arguments, such as `makePromise<T>(fn: () => T): Promise<T>`. Here while `Promise<void>` and `fn = (): void => {...}` would be allowed, `makePromise<void>(...)` is not allowed.

Hence my question: **Should `void` be allowed in generic arguments to function calls?**

It seems to have been explicitly disallowed here by including CallExpression in `invalidGrandParents`:

https://github.com/typescript-eslint/typescript-eslint/blob/407a37430d5a8ff59e111993b2585ee162fef821/packages/eslint-plugin/src/rules/no-invalid-void-type.ts#L71

https://github.com/typescript-eslint/typescript-eslint/blob/407a37430d5a8ff59e111993b2585ee162fef821/packages/eslint-plugin/tests/rules/no-invalid-void-type.test.ts#L240

Should the rule allow this by default, or should there be a rule configuration option to allow this?

**Repro**

https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/no-invalid-void-type.ts#L71

```JSON
{
  "rules": {
    "@typescript-eslint/no-invalid-void-type": "error"
  }
}
```

```TS
declare function makePromise<T>(fn: () => T): Promise<T>; // No error
makePromise<void>(()=>{}) // Error
```

**Comment by bradzacher** [comment]
The reason it's not allowed by default is that it is rare that you actually want this.
I'd hazard a guess that this case of "a function that accepts a callback function and is specifically parameterised around return type of the passed callback function" is one of the very few cases where it's going to be used.

The overwhelming majority of times when someone passes `void` into a function call's generic parameter they actually meant `undefined` instead, because they aren't typing a callback function's return value.

The intersection of people that use that rule and people that use that pattern is likely very low, evidenced by the fact that it's been a year without anyone asking for this.

I'd be happy for an option akin to the [`allowInGenericTypeArguments`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-invalid-void-type.md#allowingenerictypearguments) option - as in it accepts an array of names to allow.

I don't think having a `true` option makes as much sense there, because like I said - it's rare. If you were to turn the option to `true`, then really the rule wouldn't do much of anything any more.

**Comment by jtbandes** [comment]
> I'd be happy for an option akin to the allowInGenericTypeArguments option - as in it accepts an array of names to allow.

What if, rather than adding a new option, the rule were changed so that this `allowInGenericTypeArguments` list of allowed names is also used for function names?

**Comment by bradzacher** [comment]
best to not conflate the two.
declaring the type `T` in the list should not also allow the function `T()`.

**Comment by Xenya0815** [comment]
I ran into the same problem
An option of the rule to accept `void` in such cases would help.

**Comment by NickHeiner** [comment]
I was surprised to see [in this playground](https://typescript-eslint.io/play/#ts=4.5.2&sourceType=module&code=CYUwxgNghgTiAEAzArgOzAFwJYHtXwAcYcBbLAZy0QE8AeAFQCURzkIMA+ACgEoAueAAViZciAbNW7DgG4AUHKKkKVOgDccWYNx4ygA&rules=N4IgAgLgngDgpgZwMYCcCWMIFpEBs0B2EA9AQPZaEBuAhvgCZZVlqPTwgBcA2iHCijIoQAGlB1cZAO4BJAgHE4BfmiQAVWHACCKAOYBXALZKICLhBT64AXwC61oA&tsConfig=N4XyA) that a generic type arg was flagged as a violation, despite setting `allowInGenericTypeArguments`. Perhaps I'm misunderstanding the intent of the rule, or the option? (Also, apologies if this thread is unrelated to what I'm saying here and I should have made this its own issue.)

**Comment by trajano** [comment]
When you're testing with `@testing-library/react-hooks ` if you're testing with a hook that has no return value (e.g. `useEffect`) the the type parameter for the return value would have to be `void`.  For now I disabled the rule to remove the error

**Comment by KuSh** [comment]
Same thing with jest Mock typing : see [playground](https://typescript-eslint.io/play/#ts=4.9.5&sourceType=module&code=CYUwxgNghgTiAEA7KBbEBnADlMCBWGALvAN4BQ8l8AlooSDAGY4ICyA9mANYA8AKvAC88KIgCeAGngBNeCAAe9RMHQjxAbQC6QtWIB8cxSGWqAYgFdEYQtXaJSFKk8QgA7gAoAdN9gBzdABcMgCUQXwA3I5OlF4+MP5B0qHwEVHwAL5pjJbWtvaMiPxSsgpKKrpaeu7UKJgQIGh0UDZ2APxBsZ5+gSFCBnzJHNxFMnqRmWRgdujEKJxcOgQzngU8AG7s1MBSLmsMVcHhQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uYDcBDRgBNaPDpWFFS6KAHc+0JpHBgAviBVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA)

**Comment by phryneas** [comment]
Using `void` as a generic argument to a function call is a very common use case in Redux Toolkit when you want to specify that a generated function should not have a certain argument (we use it mainly in `createAction`, `createAsyncThunk` and `build.endpoint` of `createApi`).

This probably doesn't seem to come up as this eslint rule generally seems not to be used a lot, but it would be really great if this could also be allowed for method call generic arguments.

Here is a [TypeScript playground](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgYygUwIYzQQWTYCAOzgF84AzKCEOAcgAF0ATAVwA8ArAZwHoYIEADYBrYDDoAoSQklw4vXnDTsM4IWjgBGAFxwM+QiVSYBUOAHdxACzjM0FYETTM4YDAE8hEDMzkpibngDAmIAYXQsaAB1GwAFT29fOABeFEjsPFCiAB4kCkE9IKgnAHMyAD4ACgAiEKMAFQ8wNBqASn9FeW7ugD0Afml5evCMmPjEn2Yq-MK4GoAjDCgasjaAbk6lBhhuAFoVFvwDqGoofxGiCNNxmGsErymqjtJpWXkulTUwDTgAJjge30hmI6Ru5isdwgrHg7keySqADcIMBmB15MhAsEQVcxlBYlCYQ8kq40iYsLgcTlkajqnUcU0Wu0tj0egMhgptrsDuwjjATmcLjjrlF8TZoTBiU9ZhA9ItlqtSBshdkRWYCdYJVLfM9JK8gA) showcasing this intended use with `createAction`.

**Comment by OpportunityLiu** [comment]
Now here is a use case in the built-in function.

https://github.com/tc39/proposal-promise-with-resolvers
